toc/rfc/rfc-draft-add-cnbs.md

```diff
@@ -0,0 +1,194 @@
+| `finalize` | Add CNB `launcher` executable and runtime configuration files to build layers |
+| `release` | Package the generated layer directories into a Cloud Foundry droplet |
+
+These executables will be invoked by the Cloud Foundry Buildpack App Lifecycle's
+builder.
+
+### Shim at Run Time
+
+At run time, the Cloud Foundry Buildpack App Lifecycle launcher will invoke the
+CNB Lifecycle `launcher` executable that was included in the droplet at build
+time. In order to support custom start commands for apps using the CNB
+buildpack shims, the CNB launcher invocation will be prepended to the app start
+command using the [`entrypoint_prefix` hook](https://github.com/cloudfoundry/buildpackapplifecycle/commit/29feb13caeff646f35585eee865c376c818fc2ea)
+in the CF Lifecycle launcher.
+
+### Service Bindings
+
+The largest anticipated breaking change from v2 to v3 buildpacks is the change
+in how service bindings are consumed by buildpacks. The v3 buildpacks expect
+service bindings to follow the [Kubernetes Service Binding specification](https://github.com/servicebinding/spec),
+which places service binding credentials on the filesystem. Cloud Foundry and
+v2 buildpacks instead use environment variables for service binding
+credentials.
+
+In order to securely place credentials on the filesystem, Diego will need the
+ability to mount in-memory filesystems (e.g. `tmpfs`) in application
+containers. Once an in-memory filesystem is available, select `VCAP_SERVICES`
+environment variables can be translated into their Kubernetes Service Binding
+equivalents by the shim orchestrator, for consumption by the v3 buildpacks.
+
+### Governance and Release
+
+A `cnb-shim-builder` repository will be added to the Application Runtime
+Interfaces working group. This repository will contain the tooling necessary
+for converting existing Paketo buildpacks into shimmed CNBs.
+
+The shimmed buildpacks will be experimental bosh releases, belonging to the
+Application Runtime Interfaces working group. The proposed list of buildpacks
+is as follows (in rough priority order):
+1. `java_native_cnb_release`
+1. `java_cnb_release`
+1. `web_servers_cnb_release`
+1. `dotnet_core_cnb_release`
+1. `nodejs_cnb_release`
+1. `python_cnb_release`
+1. `ruby_cnb_release`
+1. `procfile_cnb_release`
+1. `go_cnb_release`
+1. `nginx_cnb_release`
+1. `php_cnb_release`
+1. `rust_cnb_release`
+
+It is not required to shim all buildpacks during phase 1 of this project.
+Depending on the effort to shim buildpacks and other factors, it may make sense
+to move to native CNB support before all Paketo buildpacks are shimmed. During
+phase 2, the above bosh release repositories will be converted to unshimmed
+bosh releases of the corresponding Paketo buildpacks.
+
+The CNBs will initially be opt-in for CF Deployment operators via an
+experimental ops file. Once the buildpacks have reached a suitable level of
+maturity and stability, likely after phase 2, they will be added to the default
```

Comment:
How would the buildpack detection work when v2 and v3 buildpacks are installed both?

Reply:
> How would the buildpack detection work when v2 and v3 buildpacks are installed both?

I'd expect them to be installed under different names, so you would specific them specifically with `cf push -b v3_python_buildpack`, or rely on the buildpack ordering for detect. However, given that they will generally detect on the same things (e.g. presence of language-toolchain-specific files) I would expect whichever is first in the detection order would likely win. For backwards-compatibility we would probably want to suggest operators put the v3 buildpacks lower in the order and ask users to explicitly opt-in via `cf push -b`.

Reply:
Great, that makes sense to me about the distinct buildpack names and v2/v3 ordering for appropriate autodetection. We'll need to choose those names carefully so that we're all happy enough with them for the long term, since presumably any app manifests or `cf push` commands that select buildpacks by name will use those new names and it's generally a pain for app teams to change them later without a good reason.

Reply:
I agree with the above. Also, if the v3 buildpacks are using the Paketo stacks, then the app's stack will determine which set of buildpacks it uses.

Reply:
TOC discussion:
* resolve to provide recommendations for naming, so people can knowingly opt in

Reply:
https://github.com/cloudfoundry/community/commit/ba6fa1a04b256c2e506c7457e1665f2970bd641a